Approach.md

```diff
@@ -0,0 +1,102 @@
+
+Here's a breakdown of the engineering tradeoffs for each component:
+
+## rrockerd:
+I'll be using the rust crate [nix](https://github.com/nix-rust/nix) extensively for the low level systemcalls. Since nix is merely a safe wrapper around libc I believe this is in the spirit of the challenge.
+
+The daemon will be run entirely in memory and will not persist any state in order to survive reboots/crashes.
+
+The gRPC API is assumed to be internal only, as such there won't be any rate limiting, user action logging or abuse mitigations.
+
+Dealing with tty, ptty, terminals and what not is entirely out of scope and as such any program output will be streamed as the raw bytes written to the stdout and stderr pipes.
+Furthermore the output will be chunked on newlines and any lines longer than max gRPC packet size will be truncated.
+
+Resource limits will be implemented using separate cgroups for each
+
+Isolation will be hardcoded to fully isolate each scheduled task from other tasks and the host system. That means a separate PID, Mount, User and Network namespace for each task with no way to connect them.
+There will be no network devices attached to any tasks.
+
+Mounting is also considered out of scope and as such each task will be chrooted to a RW root that's a copy of a base image that's removed upon exit of the task.
+
+I'll make no attempt at Linux backwards compatibility and the minimum kernel version will be 5.0+.
+
+Tasks will inherit the UID/GID of the daemon and internally be mapped as root.
+
+For the gRPC implementation I'll be using the rust crate [tonic](https://github.com/hyperium/tonic) to avoid reinventing the wheel.
+
+Mutable interaction with tasks will be mutually exclusive meaning that the stop action will take a write-only lock on the task such that any concurrent stop task will wait for the first to complete and the subsequently fail as the task has already stopped.
+Read-only interactions such as query and stream can happen concurrently without issue.
+
+## rrocker-cli:
+The CLI will need to be run once per command, so scheduling multiple tasks requires multiple invocations.
+
+There'll be zero command line switches and as such it'll be hardcoded to connect to `rrockerd` running on localhost.
+
+## rrocker-lib:
+No attention will be paid to backwards compatibility of the API, meaning no versioning or abstractions.
+
+Human friendliness of the API is a distant afterthought, as such full UUIDs will be required to interact with tasks.
+
+Resource limits will be exposed 1-to-1 with the underlying cgroups API leaving it up to the `rrocker-cli` to expose it in a human friendly way.
+# Security
+Security is naturally an important aspect of the implementation however as my experience with crypto is limited I'll attempt to follow best practices and use the default configuration of OpenSSL and rustTLS wherever applicable.
```

Comment:
What about other aspects of security, like the isolation of the containerization? Will any protections be taken against container escape for example?

Reply:
I will be relying on the `clone(2)` and `pivot_root(2)` to isolate the containers so barring any bugs in my implementation I'll be relying on the security of the kernel.

However briefly mentioned in the test section I'll write some basic unit tests that verifies a containerized task is unable to see the host system PIDs, make network connections and write to unauthorized paths.

More advanced container escape attacks are outside my expertise and will be considered out of scope. 
However I'd be more than happy learn about such attacks :-)

Reply:
I think we generally ask to use cgroups in order to provide robust namespacing and resource limitation as specified by the challenge.

Reply:
As far as I'm aware cgroups only provides resource utilization constraints. Namespaces provide isolation and can be activated with the `unshare(2)` and `clone(2)` syscalls.
Since I'm aiming for level 5 with both utilization limits and isolation I'll be taking advantage of both cgroups and namespaces.